OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs

```diff
@@ -35,84 +37,72 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-			{
-				if (t.ContainsGenericParameters || t.IsAbstract)
-					continue; // skip helpers like TraitInfo<T>
-
-				if (currentNamespace != t.Namespace)
-				{
-					currentNamespace = t.Namespace;
-					doc.AppendLine();
-					doc.AppendLine($"## {currentNamespace}");
-				}
-
-				var traitName = t.Name.EndsWith("Info") ? t.Name.Substring(0, t.Name.Length - 4) : t.Name;
-				var traitDescLines = t.GetCustomAttributes<DescAttribute>(false).SelectMany(d => d.Lines);
-				doc.AppendLine();
-				doc.AppendLine($"### {traitName}");
-				foreach (var line in traitDescLines)
-					doc.AppendLine(line);
-
-				var requires = RequiredTraitTypes(t);
-				var reqCount = requires.Length;
-				if (reqCount > 0)
-				{
-					if (t.HasAttribute<DescAttribute>())
-						doc.AppendLine();
-
-					doc.Append($"Requires trait{(reqCount > 1 ? "s" : "")}: ");
-
-					var i = 0;
-					foreach (var require in requires)
-					{
-						var n = require.Name;
-						var name = n.EndsWith("Info") ? n.Remove(n.Length - 4, 4) : n;
-						doc.Append($"[`{name}`](#{name.ToLowerInvariant()}){(i + 1 == reqCount ? ".\n" : ", ")}");
-						i++;
-					}
-				}
+			var json = GenerateJson(version, traitInfos, objectCreator);
+			Console.WriteLine(json);
+		}
 
-				var infos = FieldLoader.GetTypeLoadInfo(t);
-				if (!infos.Any())
-					continue;
-				doc.AppendLine();
-				doc.AppendLine("| Property | Default Value | Type | Description |");
-				doc.AppendLine("| -------- | --------------| ---- | ----------- |");
-				var liveTraitInfo = Game.ModData.ObjectCreator.CreateBasic(t);
-				foreach (var info in infos)
+		static string GenerateJson(string version, IEnumerable<Type> traitTypes, ObjectCreator objectCreator)
+		{
+			var traitTypesInfo = traitTypes.Where(x => !x.ContainsGenericParameters && !x.IsAbstract)
+				.Select(type => new
 				{
-					var fieldDescLines = info.Field.GetCustomAttributes<DescAttribute>(true).SelectMany(d => d.Lines);
-					var fieldType = Util.FriendlyTypeName(info.Field.FieldType);
-					var loadInfo = info.Field.GetCustomAttributes<FieldLoader.SerializeAttribute>(true).FirstOrDefault();
-					var defaultValue = loadInfo != null && loadInfo.Required ? "*(required)*" : FieldSaver.SaveField(liveTraitInfo, info.Field.Name).Value.Value;
-					doc.Append($"| {info.YamlName} | {defaultValue} | {fieldType} | ");
-					foreach (var line in fieldDescLines)
-						doc.Append(line + " ");
-					doc.AppendLine("|");
-				}
-			}
+					type.Namespace,
+					Name = type.Name.EndsWith("Info") ? type.Name.Substring(0, type.Name.Length - 4) : type.Name,
+					Description = string.Join(" ", type.GetCustomAttributes<DescAttribute>(false).SelectMany(d => d.Lines)),
+					RequiresTraits = RequiredTraitTypes(type)
+						.Select(y => y.Name),
+					InheritedTypes = type.BaseTypes()
+						.Select(y => y.Name)
+						.Where(y => y != type.Name && y != $"{type.Name}Info" && y != "Object" && y != "TraitInfo`1"),
```

Comment:
```y != "TraitInfo`1"``` feels like such a massive hack

Reply:
Yes.

Reply:
Maybe add a comment explaining that it's a hack and why it is needed

Reply:
ping